app/src/main/java/org/schabi/newpipe/local/playlist/LocalPlaylistFragment.java

```diff
@@ -331,6 +358,121 @@ public void onComplete() { }
+
+        removeWatchedDisposable = Flowable.just(Flowable.just(removePartiallyWatched,
+                playlistManager.getPlaylistStreams(playlistId).blockingFirst()))
+                .subscribeOn(Schedulers.io())
+                .map(flow -> {
+                    boolean localRemovePartiallyWatched = (boolean) flow.blockingFirst();
+                    List<PlaylistStreamEntry> playlist
+                            = (List<PlaylistStreamEntry>) flow.blockingLast();
+                    HistoryRecordManager recordManager = new HistoryRecordManager(getContext());
+                    Iterator<PlaylistStreamEntry> playlistIter = playlist.iterator();
+                    Iterator<StreamHistoryEntry> historyIter = recordManager
+                            .getStreamHistorySortedById().blockingFirst().iterator();
+                    List<PlaylistStreamEntry> notWatchedItems = new ArrayList<>();
+                    Iterator<StreamStateEntity> streamStatesIter = null;
+                    boolean thumbnailVideoRemoved = false;
+
+                    if (!localRemovePartiallyWatched) {
+                        streamStatesIter = recordManager.loadLocalStreamStateBatch(playlist)
+                                .blockingGet().iterator();
+                    }
+
+                    // already sorted by ^ getStreamHistorySortedById(), binary search can be used
+                    ArrayList<Long> historyStreamIds = new ArrayList<>();
+                    while (historyIter.hasNext()) {
+                        historyStreamIds.add(historyIter.next().getStreamId());
+                    }
+
+                    if (localRemovePartiallyWatched) {
+                        while (playlistIter.hasNext()) {
+                            PlaylistStreamEntry playlistItem = playlistIter.next();
+                            int indexInHistory = Collections.binarySearch(historyStreamIds,
+                                    playlistItem.getStreamId());
+
+                            if (indexInHistory < 0) {
+                                notWatchedItems.add(playlistItem);
+                            } else if (!thumbnailVideoRemoved
+                                    && playlistManager.getPlaylistThumbnail(playlistId)
+                                    .equals(playlistItem.getStreamEntity().getThumbnailUrl())) {
+                                thumbnailVideoRemoved = true;
+                            }
+                        }
+                    } else {
+                        boolean hasState = false;
+                        while (playlistIter.hasNext()) {
+                            PlaylistStreamEntry playlistItem = playlistIter.next();
+                            int indexInHistory = Collections.binarySearch(historyStreamIds,
+                                    playlistItem.getStreamId());
+
+                            hasState = streamStatesIter.next() != null;
+                            if (indexInHistory < 0 ||  hasState) {
```

Comment:
Anyway, I don't think here another variable is needed, you could just go with
```suggestion
                            if (indexInHistory < 0 || streamStatesIter.next() != null) {
```

Reply:
streamStatesIter Has to be incremented each time to keep up with the playlistIter
streamStatesIter.next() Doesn't always get called in the "OR" conditional, thus would make them out of sync, (each time) when indexInHistory >= 0

Reply:
Oh, I see, you are right